### DISCUSSION TITLE
Dynamically generated properties / subs

### DISCUSSION THREAD

**Discussion by didster** [Camelcade/Perl5-IDEA#2416] [Ideas]
Love the plugin!

Is there anyway to manually tell plugin that certain properties exist in a package ?

For example, I make heavy use of the Class::DBI module in many of my Perl programs for database ORM.   This module works by dynamically generating sub's on the fly for the database columns, none of which show in Intellisense given they dont exist as static sub's within the derived module.  I'd love to be able to say something like 

package MyPackage;

#@subs foo bar biff

And foo bar and biff are shown for Intellisense for anything that has #@type MyPackage

**Comment by hurricup** [comment]
For now you may only generate fake pm file with all declarations. but this looks like a good idea to generate subs based on db.

**Reply by hurricup** [reply]
Can say from start it will only work in paid products because they providing DB support. Not ready to write one myself.

**Reply by didster** [reply]
I dont think anything as fancy as that is needed, at least for my personal stuff.  It would just be handy to have a way of saying sub xyz exists to the IDE so it can auto complete it, even the sub has been generated at run time.

Databases are the most common example I guess but I've seen plenty of other stuff to generating things dynamically such as logging frameworks with things like

```
foreach my $l (
        qw( debug info notice warn warning err error crit critical alert emerg emergency )
        ) {
        my $sub = sub {
            my $self = shift;
            $self->log( level => $l, message => "@_" );
        };
        no strict 'refs';
        *{$l} = $sub;
}
```

It would be handy to be able to just drop #@ subs debug info warn etc into the above module so the IDE can auto complete. It would be even better to be able to say #@ subs Logging::Module debug info warn  so you could declare the runtime subs outside of the module itself if you were not the modules author.
